src/core/namespace.c

```diff
@@ -1413,15 +1413,39 @@ static int mount_image(
 }
 
 static int mount_overlay(const MountEntry *m) {
-        const char *options;
+        _cleanup_free_ char *options = NULL, *dirs = NULL;
+        _cleanup_strv_free_ char **overlays = NULL;
         int r;
 
         assert(m);
 
-        options = strjoina("lowerdir=", mount_entry_options(m));
-
         (void) mkdir_p_label(mount_entry_path(m), 0755);
 
+        /* Extension hierarchies are optional (e.g.: confext might not have /opt) so check if they actually
+         * exist in an image before attempting to create an overlay with them, otherwise the mount will
+         * fail. We can't check before this, as the images will not be mounted until now. */
+
+        overlays = strv_split(mount_entry_options(m), ":");
```

Comment:
Hmm, I am only seeing this right now, that for MOUNT_OVERLAY entries we store the concatenated list of lowerdirs in the options string? yikes.

I am not a fan of storing stuff as text internally, and then trying to parse this again. Here even escaping is applied: when the option string was put together the ":" and "," characters needed to be escaped, and thus if you split things up here and want to use it as a regular path you have to unescape again.

I am also not a fan of naming a field "options", and then putting something in there that actually isn't the options. but still needs to be prefixed with "lowerdir="...

I understand this is a pre-existing thing, but I am not happy with just piling on.

So I think append_extensions() should not try do escape/generate the lowerdir= string, but instead just store the relevant paths in a strv, one per hierarchy in MountEntry. And then here in mount_overlay() we go through that, suppress missing dirs, escape, concatenate and mount this stuff.

Reply:
ok, updated